[조이] Dev/sprint01 - Midpoint

youzysu opened the pull request:
# 주요 작업

- 개발 환경 초기 세팅
- 최신 뉴스, 언론사 브랜드 마크 데이터 크롤링
- Component 클래스 
- 기본 상단 영역 UI
- [최신 뉴스 자동 롤링 영역 작업중](https://github.com/youzysu/fe-max--newsstand/issues/3)
  - 상위 컴포넌트에서 전체 뉴스 데이터 fetch 후 state에 저장
  - 각 newsBar에 props로 전달 예정

youzysu committed:
chore: 프론트엔드 빌드 개발 환경 초기 세팅

- 프론트엔드 개발: TypeScript, module.css
- 빌드 도구: Vite

youzysu committed:
style, chore: 공통 스타일 파일 생성 및 설정

youzysu committed:
feat #1 #2: Base Component Class 생성 후 기본 상단 영역 UI 생성

- Header Component 생성
- 하위 Title, TodaysDate 컴포넌트 생성

youzysu committed:
feat #7 #8: 네이버 뉴스스탠드 데이터 크롤링하여 JSON 파일 생성

- pupeteer 라이브러리 활용

youzysu committed:
feat: express server 설정

- trend news, media endpoint 설정
- cors module

youzysu committed:
feat #6: 최신 뉴스 자동 롤링 영역 데이터 받아오기

- 상위 컴포넌트에서 전체 뉴스 데이터 fetch 후 state에 저장하기
- 각 newsBar에 props로 전달 예정

youzysu committed:
chore: 불필요한 파일 삭제

youzysu committed:
feat #6: 최신 뉴스 영역 컴포넌트 생성

- 상위 컴포넌트에서 전체 뉴스 데이터를 받아 각 NewsBar 컴포넌트에 5개씩 전달한다.

youzysu committed:
fix: express 오류로 port 변경

hjsong333 commented on src/components/Header/index.ts:
```diff
@@ -0,0 +1,46 @@
+      month: '2-digit',
+      day: '2-digit',
+      weekday: 'long',
+    }).format(this.state.currentTime);
```
Intl 이게 뭔지 몰랐는데, 빌트인 생성자였군요. 좋은거 배워갑니다.

hjsong333 commented on package.json:
```diff
@@ -0,0 +1,34 @@
+    "express": "^4.18.2",
+    "nodemon": "^2.0.22",
+    "puppeteer": "^20.3.0",
+    "ts-node": "^10.9.1"
```
개발시에 필요한거라 devDependency 쪽으로 옮겨야할 것 같습니다.

hjsong333 commented on src/utils/createElement.ts:
```diff
@@ -0,0 +1,12 @@
+  [name: string]: string;
+}
+
+export const createElement = (tagName: string, attributes?: attributes): HTMLElement => {
+  const element = document.createElement(tagName);
+  if (attributes) {
+    Object.keys(attributes).forEach((name) => element.setAttribute(name, attributes[name]));
+  }
+
+  return element;
+};
```
template literal 대신 함수를 만드신 이유가 있을까요?

hjsong333 commented on src/components/base/index.ts:
```diff
@@ -0,0 +1,55 @@
+
+  addEvent(eventType: string, selector: string, handler) {
+    this.element.addEventListener(eventType, (event) => {
+      if (!event.target?.closest(selector)) {
```
closest 이걸 꼭 사용해야할까요?

hjsong333 commented on src/components/NewsStand.ts:
```diff
@@ -0,0 +1,18 @@
+import Header from './Header';
+import TrendNews from './TrendNews';
+import Component from './base';
+import styles from './newsstand.module.css';
```
이거 너무 편해보이네요.

hjsong333 commented on src/server/data/crawler.mjs:
```diff
@@ -0,0 +1,61 @@
+  const allNewsMedia = [];
+
+  while (allNewsMedia.length < 4) {
+    const newsMedia = await page.$$eval(newsMediaSelector, (elements) =>
```
querySelectorAll 보다 간결해서 편안하네요

hjsong333 reviewed (commented):
진행을 많이 하셨는데...보고 배울점이 많은 코드였습니다. 처음 봤을 때 살짝 헷갈리는 부분이 있어서 당황했는데, 그 부분 빼고는 제 스타일입니다.

youzysu committed:
chore: dependencies, devDependencies 수정

- 페어 리뷰 반영

youzysu committed:
refactor: State type 수정

youzysu committed:
chore: tsconfig.json 파일 오류 수정

youzysu committed:
refactor #1: 컴포넌트 클래스 구조 설계 변경 및 Store 분리

- 초기화 시 자식 컴포넌트, HTML Element를 property로 가진다.
- Store에서 모든 상태를 관리한다.
- 각 컴포넌트는 렌더링을 위해 필요한 상태를 상위 컴포넌트에서 props로 전달 받는다.

youzysu committed:
refactor #6: NewsBar AutoScrollingNews 컴포넌트 UI 로직 수정

- 왼쪽 뉴스 리스트와 오른쪽 뉴스 리스트 분리하여 props 전달
- currentIndex, nextIndex에 해당하는 뉴스 헤드라인 생성 & wrapper

youzysu committed:
feat, refactor #9: NewsBar 5초마다 자동으로 뉴스가 무한 롤링된다.

- dispatch 함수로 action을 전달하여 reducer로 상태를 변경한다.
- 해당 상태를 subscribe하는 함수에 변경된 상태를 인자로 전달하여 실행한다.
- 변경된 새로운 props를 전달받은 컴포넌트는 현재 상태와 비교하여 변경된 부분만 반영한다.
- refactor: 네이밍 일관성을 위해 media를 press로 변경

youzysu committed:
feat #10: Tab Viewer 기본 UI 생성 및 Option State에 따른 active styling

youzysu committed:
chore: 디렉터리 구조 수정

- server 디렉터리 최상단으로 변경

youzysu committed:
feat #3: 뉴스 리스트 inital state fetch 하기

youzysu committed:
feat #5: AllPressList 전체 데이터 fetch 후 24개씩 Chunking하여 저장한다.

youzysu committed:
feat #9: AutoRollingNewsBar Animation 구현

- rAF callback으로 5초 후에 rolling class 추가하기
- transitionend 이벤트 시 다음 news UI로 변경 및 rolling class 제거하기
- refactor: 매직 넘버 수정

youzysu committed:
feat #5: 6 * 4 테이블로 구성된 GridViewer UI 생성

- allpressList 전체 96개의 press data를 fetch 받는다.
- allpressList와 현재 페이지의 gridPressStartIndex를 인자로 받아 테이블을 구성한다.
- GridViewer 컴포넌트는 table 태그, 각 Grid는 td 태그, 4개의 row로 이루어진다.
